jerry-core/ecma/builtin-objects/ecma-builtin-helpers-sort.c

```diff
@@ -17,133 +17,137 @@
     }
 
-    JERRY_ASSERT (child <= right);
-
-    /* Compare current child node with the swap (tree top). */
-    ecma_value_t swap_compare_value = sort_cb (array_p[child], swap, compare_func);
-
-    if (ECMA_IS_VALUE_ERROR (swap_compare_value))
+    if (i < right_idx && ecma_get_number_from_value (compare_value) <= ECMA_NUMBER_ZERO)
     {
-      ret_value = ECMA_VALUE_ERROR;
-      break;
+      array_b_p[k] = array_a_p[i];
+      i++;
     }
-
-    JERRY_ASSERT (ecma_is_value_number (swap_compare_value));
-
-    if (ecma_get_number_from_value (swap_compare_value) <= ECMA_NUMBER_ZERO)
+    else
     {
-      /* Break from loop if current child is less than swap (tree top) */
-      ecma_free_value (swap_compare_value);
-      break;
+      array_b_p[k] = array_a_p[j];
+      j++;
     }
-
-    /* We have to move 'swap' lower in the tree, so shift current child up in the hierarchy. */
-    uint32_t parent = (child - 1) / 2;
-    JERRY_ASSERT (parent <= right);
-    array_p[parent] = array_p[child];
-
-    /* Update child to be the left child of the current node. */
-    child = child * 2 + 1;
-
-    ecma_free_value (swap_compare_value);
+    ecma_free_value (compare_value);
   }
 
-  /*
-   * Loop ended, either current child does not exist, or is less than swap.
-   * This means that 'swap' should be placed in the parent node.
-   */
-  uint32_t parent = (child - 1) / 2;
-  JERRY_ASSERT (parent <= right);
-  array_p[parent] = swap;
-
   return ret_value;
-} /* ecma_builtin_helper_array_to_heap */
+} /* ecma_builtin_helper_array_merge_sort_bottom_up */
 
 /**
- * Heapsort function
+ * Mergesort function
  *
  * @return ecma value
  *         Returned value must be freed with ecma_free_value.
  */
 ecma_value_t
-ecma_builtin_helper_array_heap_sort_helper (ecma_value_t *array_p, /**< array to sort */
-                                            uint32_t right, /**< right index */
-                                            ecma_value_t compare_func, /**< compare function */
-                                            const ecma_builtin_helper_sort_compare_fn_t sort_cb) /**< sorting cb */
+ecma_builtin_helper_array_merge_sort_helper (ecma_value_t *array_p, /**< array to sort */
+                                             uint32_t length, /**< length */
+                                             ecma_value_t compare_func, /**< compare function */
+                                             const ecma_builtin_helper_sort_compare_fn_t sort_cb) /**< sorting cb */
 {
-  /* First, construct the ordered binary tree from the array. */
-  for (uint32_t i = (right / 2) + 1; i > 0; i--)
+  // create array copy
+  ecma_value_t ret_value = ECMA_VALUE_EMPTY;
+  JMEM_DEFINE_LOCAL_ARRAY (array_copy_p, length, ecma_value_t);
+
+  uint32_t index = 0;
+
+  while (index < length)
   {
-    ecma_value_t value = ecma_builtin_helper_array_to_heap (array_p, i - 1, right, compare_func, sort_cb);
+    array_copy_p[index] = array_p[index];
```

Comment:
Do we need to initialize this array? The first iteration will overwrite it, and if there are no iterations, then we would just free it anyway.